### PR TITLE
Show layers when pulling image

### PR DESCRIPTION
**Pull request by theMagicalKarp**
screen shot 2015-04-11 at 1 52 16 pm

### CODE REVIEW

File: app/js/instances/instanceModel.js [hunk 1]
```diff
@@ -118,29 +140,9 @@ function instanceModel(dockerService) {
...
-            }
-
-            if (_.has(r.response, 'progressDetail') && _.has(r.response.progressDetail, 'current')) {
-                var progress = r.response.progressDetail;
-                this.loadingImages[imageTag] = 100 * progress.current / progress.total + '%';
-                this.emitChange();
             }
         },
 
-        removeImage: function(r) {
```
Comment (ngmiller) [inline app/js/instances/instanceModel.js]: Why was this removed?

File: app/js/instances/instanceDetailController.js [hunk 2]
```diff
@@ -50,12 +50,7 @@ function instanceDetailController($scope, $routeParams, $timeout, flux, dockerSe
...
-
-        var loadingImages = instanceModel.getLoadingImages();
-        _.forEach(loadingImages, function(progress, name) {
-            $scope.loadingImages[name] = progress;
-        });
-        $scope.loadingImages = _.pick($scope.loadingImages, _.keys(loadingImages));
+        $scope.loadingImages = instanceModel.getLoadingImages();
 
         // this needs some investigating, without this, loadingImages won't show
         $timeout(function() {
```
Comment (ngmiller) [inline app/js/instances/instanceDetailController.js]: Does this work without the timeout now?

File: app/js/instances/templates/instance.html [hunk 3]
```diff
@@ -69,11 +69,14 @@
...
-    <div class="col-md-12" ng-repeat="(name, progress) in loadingImages">
-      <span>Pulling <strong>{{ name }}</strong></span>
-      <div class="progress">
-        <div class="progress-bar progress-bar-info" role="progressbar" aria-valuemin="0" aria-valuemax="100" ng-style="{'width': progress}"></div>
-      </div>
+    <div class="col-md-12" ng-repeat="(name, layers) in loadingImages track by name">
+      <span>Pulling <strong>{{ name }}</strong> <i class="fa fa-spinner fa-spin"></i></span>
+      <div ng-repeat="layer in layers track by layer.id">
+        <div class="progress">
+          <div class=" progress-bar progress-bar-info" role="progressbar" aria-valuemin="0" aria-valuemax="100" ng-style="{'width': layer.progress}">
```
Comment (ngmiller) [inline app/js/instances/templates/instance.html]: From a UX perspective, the centered progress bar text may get a little annoying as the bar grows. Could we push it off to the top or left of the progress bar?

File: app/js/instances/instanceDetailController.js (hunk 2 above)
Comment (theMagicalKarp) [inline app/js/instances/instanceDetailController.js]: I tried but it still doesn't work without the timeout

File: app/js/instances/instanceModel.js (hunk 1 above)
Comment (theMagicalKarp) [inline app/js/instances/instanceModel.js]: whoops

**Comment by zach-taylor** [comment]
@theMagicalKarp I'm getting this when pulling django:latest : `Uncaught Error: Cannot call dispatch while another dispatch is executing. Attempted to execute 'alertCreate' but 'pullImage' is already executing.` from `flux-angular.min.js:1`

**Comment by theMagicalKarp** [comment]
Weird, I can not reproduce.  I'll have to checkout what's going on during our meeting today.

File: app/js/instances/instanceModel.js [hunk 4]
```diff
@@ -103,12 +103,34 @@ function instanceModel(dockerService) {
                 });
```
Comment (ngmiller) [inline app/js/instances/instanceModel.js]: So this `alertService.create()` is what's causing @zach-taylor's issue with the django pull. The reason this won't work but the `dockerService.d('image.list', ...)` will is that the docker service has a noticeable round-trip time - long enough for the initial flux action to finish dispatching before the result comes back in. But, since the alerts are local, it interrupts the call stack.

One option is to set some kind of state flag in the model here so the listener can pick up on it, and then fire off the alert.
